File: two-pointers/680/680.js
/**
 * @param {string} s
 * @return {boolean}
 */
var validPalindrome = function (s) {
  let [endLeft, endRight] = returnTwoPointers(s, 0, s.length - 1);
  if (endLeft >= endRight) {
    return true;
  }
  return (
    isPalindrome(s, endLeft + 1, endRight) ||
    isPalindrome(s, endLeft, endRight - 1)
  );
};

const isPalindrome = (s, left, right) => {
  const [endLeft, endRight] = returnTwoPointers(s, left, right);
  return endLeft >= endRight;
};

const returnTwoPointers = (s, left, right) => {
  while (left < right) {
    if (s[left] !== s[right]) {
      return [left, right];
    }
    left++;
    right--;
  }
  return [left, right];
};

File: two-pointers/680/680.py
# https://leetcode.com/problems/valid-palindrome-ii/


class Solution:
    def validPalindrome(self, s: str) -> bool:
        left, right = self.two_pointers(s, 0, len(s) - 1)
        if left >= right:
            return True
        return self.is_palindrome(s, left + 1, right) or self.is_palindrome(
            s, left, right - 1
        )

    def is_palindrome(self, s, left, right):
        left, right = self.two_pointers(s, left, right)
        # "bddb" is a palindrome and at the end left > right
        return left >= right

    def two_pointers(self, s, left, right):
        while left < right:
            if s[left] != s[right]:
                return left, right
            left += 1
            right -= 1
        return left, right
